File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    let result = 0
    const visited = Array(maps.length).fill(0).map(()=> Array(maps[0].length).fill(0))
    const dy = [1, 0, -1, 0]
    const dx = [0, 1, 0, -1]
    const q = []
    q.push([0, 0])
    visited[0][0] = 1
    
    if (maps[maps.length - 1][maps[0].length - 2] === 0 && maps[maps.length - 2][maps[0].length - 1] === 0) return -1
    
    while (q.length) {
       const [y, x] = q.shift()
        
        for (let i = 0; i < 4; i++) {
            let ny = y + dy[i]
            let nx = x + dx[i]
            
            if (ny < 0 || nx < 0 || ny >= maps.length || nx >= maps[0].length || maps[ny][nx] === 0 ) continue
            if (visited[ny][nx]) continue
            q.push([ny, nx])
            visited[ny][nx] = visited[y][x] + 1
        }
    }
    
    result = visited[maps.length - 1][maps[0].length - 1]
    
    if(!result) return -1
    return result
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    rows = len(maps)
    cols = len(maps[0])
    
    visited = [[0] * cols for _ in range(rows)]
    
    dy = [1, 0, -1, 0]
    dx = [0, 1, 0, -1]
    
    q = deque()
    
    q.append((0, 0))
    visited[0][0] = 1
    
    if rows >= 2 and cols >= 2:
        if maps[rows - 1][cols - 2] == 0 and maps[rows - 2][cols - 1] == 0:
            pass
    while q:
        y, x = q.popleft()
        
        for i in range(4):
            ny, nx = y + dy[i], x + dx[i]
            
            if 0 <= ny < rows and 0 <= nx < cols:
                if maps[ny][nx] == 1 and visited[ny][nx] == 0:
                    visited[ny][nx] = visited[y][x] + 1
                    q.append((ny, nx))

    result = visited[rows - 1][cols - 1]
    
    if result == 0:
        return -1
    else:
        return result